client/ext/bigquery/current-timestamp-optimizer.js

```diff
@@ -0,0 +1,191 @@
+/**
+ * @copyright Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @fileoverview The CurrentTimestampOptimizer replaces the non-cacheable
+ *    CURRENT_TIMESTAMP() in queries with a fixed datetime based on the specified
+ *     granularity.
+ * 
+ * Granularity is used to provide an estimate of the current date/time, rounded to the nearest
+ * previous 'x'.  An example of this can be seen below:
+ *     original    CURRENT_TIMESTAMP()  2016-02-13 14:43:26
+ *     granularity YEAR                 2016-01-01
+ *                 MONTH                2016-02-01
+ *                 WEEK                 2016-02-06 (Saturday of previous week)
+ *                 DAY                  2016-02-13
+ *                 HOUR                 2016-02-13 14:00:00
+ *
+ * @author joemu@google.com (Joe Allan Muharsky)
+ */
+
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampGranularity');
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampOptimizer');
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampOptimizerConfigModel');
+
+
+goog.scope(function() {
+  const bigquery = p3rf.perfkit.explorer.ext.bigquery;
+  
+  const DATE_ROUNDING_TIMEZONE_OFFSET = 0;
```

Comment:
Please add the unit (minutes?) to the constant name. Or at least a comment.

Also, the direction of the offset is unobvious.

Alternatively, how about this:

```
// Number of hours after midnight in UTC to which dates should be rounded.
// For example, set this to 8 if day-granularity dates should be rounded to the
// most recent 08:00:00 in UTC. This is currently a constant, in the
// future it's planned to be replaced with a server-supplied offset.
const DATE_ROUNDING_UTC_HOUR = 0;
```

Then below, instead of setTimezoneOffset, do this:

```
var now = date.getTime();
date.setUTCHours(DATE_ROUNDING_UTC_HOUR);
// If this puts us in the future, subtract a day.
if (date.getTime() > now) {
  date = new Date(date.getTime() - 86400*1000);
}
```